Fix race and deadlock in SMP context switch paths

andyross opened the pull request:
Two bugs found by one exerciser in #28105 that repeatedly hammered the thread preemption paths.

The first was simple enough, some needless defensive coding (not wanting to change a preexisting switch handle) could be seen by code spinning in wait_for_switch() as a completed context switch and cause the other to start running the current thread before it had spilled its context.

The second was an even rarer (and more interesting) deadlock exposed when that was fixed, in a similar area of code.  Crazy timing (possible seemingly only under qemu) could lead to two CPUs trying to run **each other's** active threads, and spinning forever waiting for the other CPU to context switch.

Fixes #28105

Commit:
kernel/sched: Fix race with switch handle

The "null out the switch handle and put it back" code in the swap
implementation is a holdover from some defensive coding (not wanting
to break the case where we picked our current thread), but it hides a
subtle SMP race: when that field goes NULL, another CPU that may have
selected that thread (which is to say, our current thread) as its next
to run will be spinning on that to detect when the field goes
non-NULL.  So it will get the signal to move on when we revert the
value, when clearly we are still running on the stack!

In practice this was found on x86 which poisons the switch context
such that it crashes instantly.

Instead, be firm about state and always set the switch handle of a
currently running thread to NULL immediately before it starts running:
right before entering arch_switch() and symmetrically on the interrupt
exit path.

Fixes #28105

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Refactor, unify management of QUEUED state

The QUEUED state flag was managed separately from the run queue
insertion/deletion, and the logic (while AFAICT perfectly correct) was
tangled in a few places trying to keep them in sync.  Put the
management of both behind a queue_thread()/dequeue_thread() API for
clarity.  The ALWAYS_INLINE usage seems to be working to get the
compiler to condense the resulting multiple assignments.  No behavior
change.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

Commit:
kernel/sched: Fix rare SMP deadlock

It was possible with pathological timing (see below) for the scheduler
to pick a cycle of threads on each CPU and enter the context switch
path on all of them simultaneously.

Example:
   * CPU0 is idle, CPU1 is running thread A
   * CPU1 makes high priority thread B runnable
   * CPU1 reaches a schedule point (or returns from an interrupt) and
     decides to run thread B instead
   * CPU0 simultaneously takes its IPI and returns, selecting thread A

Now both CPUs enter wait_for_switch() to spin, waiting for the context
switch code on the other thread to finish and mark the thread
runnable.  So we have a deadlock, each CPU is spinning waiting for the
other!

Actually, in practice this seems not to happen on existing hardware
platforms, it's only exercisable in emulation.  The reason is that the
hardware IPI time is much faster than the software paths required to
reach a schedule point or interrupt exit, so CPU1 always selects the
newly scheduled thread and no deadlock appears.  I tried for a bit to
make this happen with a cycle of three threads, but it's complicated
to get right and I still couldn't get the timing to hit correctly.  In
qemu, though, the IPI is implemented as a Unix signal sent to the
thread running the other CPU, which is far slower and opens the window
to see this happen.

The solution is simple enough: don't store the _current thread in the
run queue until we are on the tail end of the context switch path,
after wait_for_switch() and going to reach the end in guaranteed time.

Note that this requires changing a little logic to handle the yield
case: because we can no longer rely on _current's position in the run
queue to suppress it, we need to do the priority comparison directly
based on the existing "swap_ok" flag (which has always meant
"yielded", and maybe should be renamed).

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Pushing this before I have a working twister locally.  I'm seeing some weird failures on qemu_leon3 and qemu_cortex_a53 and need to see what's actually happening in CI.

andyross commented:
Yeah, the leon3 and a53 failures seem to be phantom and local to my installation (I ran west update, I swear!)

But the handful of remaining failures are legitimate.  The last patch in this PR breaks k_yield().  Oops.  Fix incoming.

pabigot reviewed (commented):
Nice work.  The first two commits look fine, but I was still trying to convince myself the third was correct before it was identified as having a problem.  Will update tomorrow after the planned fix gets pushed.

andyross commented:
Add the needed logic to handle yielded current when it no longer gets added to the runqueue at the yield.

andyross commented:
For those reviewing: it's admittedly true that the final patch in this series is a little subtle.  We for sure want the bug fix in the tree, but as it's just one medium bug (that doesn't actually affect real hardware due to timing constraints with IPIs) it might be worth considering pushing this until after the release and then doing a backport.

pabigot commented:
My concern with the third commit is it has this:

    if (z_is_thread_queued(old_thread)) {
             _priq_run_add(&_kernel.ready_q.runq,
                           old_thread);
    }

which is odd because if the thread's queued like it says, why do we need to queue it again?

What I see is that `queue_thread()` will flag the thread as queued without actually queuing if SMP and thread is not the current thread. Presumably this is the situation that needs to be mitigated by the conditional `_priq_run_add()` call above.

Assessing whether each call site might flag a thread without queuing it:
* Line 318 in `next_up()` won't because it passes `_current`
* 336 `move_thread_to_end_of_prio_q()` might
* 476 in `ready_thread()` might
* 868 in `z_set_prio()` would if it isn't passed `_current`
* 1291 in `z_impl_k_thread_deadline_set()` would if it isn't passed `_current`
* 1323 in `z_impl_k_yield()` won't because it passes `_current`

At least 336 and 1291 leave me worried that a thread might be marked as queued, but never actually be queued, as I dont see a path from them to `z_requeue_current()` which is the only other place flagged-without-being-queued might be fixed.  I didn't look closely at the other potential call sites.

What prevents this from leaving a thread flagged as queued but not in a queue?

andyross commented:
I swear I convinced myself the leon3 failures had disappeared, but they hadn't and the root cause was a broken arch_switch() on sparc.  Worked around here (it's simple), root cause bug linked.

@pabigot The point to the refactor in the second patch is to clarify that the run queue (_priq_run_add/remove()) and QUEUED flags get managed through the same API, and that they are identical exactly **except** for the case of _current, which remains outside the global run queue until the point of a context switch.  That happens in two spots: synchronously in kswap.h:do_swap() and via interrupt return where z_get_next_switch_handle() is called to fetch the next thread.

andyross commented:
Pfft.  ARM64 had a different bug with arch_switch().  NOW it should work, heh.

pabigot commented on kernel/sched.c:
```diff
@@ -181,6 +181,55 @@ static ALWAYS_INLINE struct k_thread *_priq_dumb_mask_best(sys_dlist_t *pq)
+ */
+static inline bool should_queue_thread(struct k_thread *th)
+{
+	return !IS_ENABLED(CONFIG_SMP) || th != _current;
```
>  **except** for the case of _current, which remains outside the global run queue until the point of a context switch.

A-hah.  I misread the condition.

    !IS_ENABLED(CONFIG_SMP) || th != _current;

does not mean "don't queue if SMP and *not* current thread.  It's equivalent to the **much** clearer expression:

    !(IS_ENABLED(CONFIG_SMP) && (th == _current));

which means "don't queue if SMP and *is* current thread".

Then it makes sense.

I wouldn't object to using the alternative expression here.

Commit:
include/kernel_arch_interface.h: Redocument arch_switch()

Some recent changes exposed some common "arch_switch() anti-patterns"
in various architectures.  The documentation technically described
this all correctly, but probably wasn't as clear as it should have
been.  Rewrite, making clear exactly what needs to happen and how the
fields should be interpreted.

Signed-off-by: Andy Ross <andrew.j.ross@intel.com>

andyross commented:
Squash the bug workaround patch into the main change, as I realized (1) that this can be implemented by just disabling that write on !SMP (thus making it a performance win and not a workaround) and (2) it broke bisectability anyway.  Also rewrite the docs on arch_switch(), as they weren't great and probably part of the problem at the arch layer.

dcpleung commented on kernel/include/kernel_arch_interface.h:
```diff
@@ -78,50 +78,54 @@ void arch_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+ *
+ * Note that on SMP systems, the kernel uses the store through the second
+ * pointer as a synchronization point to detect when a thread context is
+ * completely saved (so another CPU can no when it is safe to switch).
```
s/no/know/?

andyross replied:
Oops.  Fixed.
